src/mlpack/tests/sfinae_test.cpp

```diff
@@ -0,0 +1,124 @@
+
+template<typename Class, typename...T>
+using MForm3 = void(Class::*)(const arma::mat&, const arma::rowvec&,
+    const arma::rowvec&, T...);
+
+HAS_METHOD_FORM(M, HasM);
+HAS_METHOD_FORM(template M<arma::mat>, HasTemplatedM);
+HAS_METHOD_FORM(SINGLE_ARG(template M<arma::mat, arma::rowvec>),
+    HasVeryTemplatedM);
+
+/*
+ * Test at compile time the presence of methods of the specified forms with the
+ * stated number of additional arguments.
+ */
+BOOST_AUTO_TEST_CASE(HasMethodFormWithNAdditionalArgsTest)
+{
+  static_assert(!HasM<A, MForm1>::WithNAdditionalArgs<0>::value,
+      "value should be false");
+  static_assert(HasM<A, MForm1>::WithNAdditionalArgs<1>::value,
+      "value should be true");
+  static_assert(HasM<A, MForm1>::WithNAdditionalArgs<2>::value,
+      "value should be true");
+
+  static_assert(!HasM<B, MForm1>::WithNAdditionalArgs<0>::value,
+      "value should be false");
+  static_assert(!HasM<B, MForm1>::WithNAdditionalArgs<1>::value,
+      "value should be false");
+  static_assert(!HasM<B, MForm1>::WithNAdditionalArgs<2>::value,
+      "value should be false");
+
+  static_assert(!HasM<A, MForm2>::WithNAdditionalArgs<0>::value,
+      "value should be false");
+  static_assert(!HasM<A, MForm2>::WithNAdditionalArgs<1>::value,
+      "value should be false");
+  static_assert(!HasM<A, MForm2>::WithNAdditionalArgs<2>::value,
+      "value should be false");
+
+  static_assert(HasM<B, MForm2>::WithNAdditionalArgs<0>::value,
+      "value should be true");
+  static_assert(!HasM<B, MForm2>::WithNAdditionalArgs<1>::value,
+      "value should be false");
+  static_assert(!HasM<B, MForm2>::WithNAdditionalArgs<2>::value,
+      "value should be false");
+
+  static_assert(!HasTemplatedM<B, MForm2>::WithNAdditionalArgs<0>::value,
+      "value should be false");
+  static_assert(HasTemplatedM<B, MForm2>::WithNAdditionalArgs<1>::value,
+      "value should be true");
+  static_assert(!HasTemplatedM<B, MForm2>::WithNAdditionalArgs<2>::value,
+      "value should be false");
+
+  static_assert(!HasVeryTemplatedM<B, MForm3>::WithNAdditionalArgs<0>::value,
+      "value should be false");
+  static_assert(!HasVeryTemplatedM<B, MForm3>::WithNAdditionalArgs<1>::value,
+      "value should be false");
+  static_assert(HasVeryTemplatedM<B, MForm3>::WithNAdditionalArgs<2>::value,
+      "value should be true");
+}
+
+/*
+ * Test at compile time the presence of methods of the specified forms.
+ */
+BOOST_AUTO_TEST_CASE(HasMethodFormTest)
+{
+  static_assert(HasM<A, MForm1>::value, "value should be true");
+
+  static_assert(!HasM<B, MForm1>::value, "value should be false");
+
+  static_assert(!HasM<A, MForm2>::value, "value should be false");
+
+  static_assert(HasM<B, MForm2>::value, "value should be true");
+
+  static_assert(HasTemplatedM<B, MForm2>::value, "value should be true");
+
+  static_assert(HasVeryTemplatedM<B, MForm3>::value, "value should be true");
```

Comment:
I think you should use `BOOST_REQUIRE_EQUAL()` or `BOOST_REQUIRE()` here, so that upon failure the Boost UTF will fail in the expected way (and also so it correctly counts the successes in the test output).